README.md
Titanium Code Processor [![Build Status](https://travis-ci.org/appcelerator/titanium-code-processor.png)](https://travis-ci.org/appcelerator/titanium-code-processor)
=======================

The Titanium Code Processor is a tool for analyzing JavaScript code in
[Titanium Mobile](https://github.com/appcelerator/titanium_mobile) projects. It
provide a wide variety of useful functionality, including runtime error detection,
Titanium API deprecation warnings, platform specific API validation, and more.
It is built using a robust plugin solution that makes it relatively easy to add
new analyses to the code processor. The tool can be used as a stand-alone binary,
used as part of the Titanium CLI, or incorporated as a library into other node.js
applications (See the API documentation in the 'docs' folder for more
information on using the code processor as a library).

Note: In this document, whenever a data structure is defined, it uses the following convention:

**[{name}]** [&lt;optional&gt;] _{type}_ [{description}]

Fields in [] are optional. If the type definition does not explicitly say &lt;optional&gt;, then it is required. If a
type has children, i.e. it is an object or array, then it's children are specified using indented lines immediately
below the type definition. Object properties always have a name and array entries never have a name.

## Table of Contents
* [Quick Start](#quick-start)
	* [Install Using NPM](#install-using-npm)
	* [Run](#run)
* [Running Using the CLI](#running-using-the-cli)
	* [Command Options](#command-options)
	* [Stream Output Format](#stream-output-format)
		* [Low Level Packet Format](#low-level-packet-format)
		* [High Level Packet Format](#high-level-packet-format)
		* [Message Types](#message-types)
	* [Config File](#config-file)
* [Running as Part of a Build](#running-as-part-of-a-build)
* [Runtime Options](#runtime-options)
* [Built-in Plugins](#built-in-plugins)
* [Internal Concepts](#internal-concepts)
* [Running the ECMA-262 Unit Tests](#running-the-ecma-262-unit-tests)

## Quick Start

### Install Using NPM

```
[sudo] npm install -g titanium-code-processor
```

The code processor works as a command in the [Titanium CLI](https://github.com/appcelerator/titanium), so make sure
it is installed before continuing. If you have a 3.0.0. or newer Titanium SDK installed, you should already have the CLI.
There are two ways to include the code processor as part of the CLI:

Automatic Method: run the included install script

```
node /path/to/titanium-code-processor/bin/install
```

Manual Method: configure the cli by hand

```
titanium config paths.commands --append /path/to/titanium-code-processor/commands
titanium config paths.hooks --append /path/to/titanium-code-processor/hooks
```

Note: On *NIX systems, the code processor is typically installed in /usr/local/lib/node_modules/titanium-code-processor

### Run

From within your project directory, run:

```
titanium analyze -p iphone -A
```

## Running Using the CLI

```
titanium analyze [options]
```
Analyzes a project.

### Command Options
<table>
	<tr>
		<th>Option</th><th>Description</th>
	</tr>
	<tr>
		<td>-A, --all-plugins</td><td>loads all plugins in the default search path</td>
	</tr>
	<tr>
		<td>--exact-mode</td><td>enables exact mode evaluation. Exact mode does not use ambiguous modes and throws an exception if an Unknown type is encountered (ignored if --config-file is specified)  [default: false]</td>
	</tr>
	<tr>
		<td>--no-console-passthrough</td><td>Prevents console.* calls in a project from being logged to the console (ignored if --config-file is specified)  [default: false]</td>
	</tr>
	<tr>
		<td>--no-loop-evaluation</td><td>Whether or not to evaluate loops (ignored if --config-file is specified)  [default: false]</td>
	</tr>
	<tr>
		<td>--no-method-invokation</td><td>prevents methods from being invoked (ignored if --config-file is specified)  [default: false]</td>
	</tr>
	<tr>
		<td>--no-native-exception-recovery</td><td>disables recovering from native exceptions when not in try/catch statements (ignored if --config-file is specified)  [default: false]</td>
	</tr>
	<tr>
		<td>--process-unvisited-code</td><td>when set to true, all nodes and files that are not visited/skipped will be processed in ambiguous mode after all other code has been processed. While this will cause more of a project to be analyzed, this will decrease accuracy and can generate a lot of false positives (ignored if --config-file is specified)  [default: false]</td>
	</tr>
	<tr>
		<td>-F, --config-file [value]</td><td>the path to the config file, note: most options and flags are ignored with this option
	</tr>
	<tr>
		<td>--cycle-detection-stack-size [size]</td><td>the size of the cycle detection stack. Cycles that are larger than this size will not be caught  [default: 10000]
	</tr>
	<tr>
		<td>--execution-time-limit [time limit]</td><td>the maximum time the app is allowed to run before erroring. 0 means no time limit (ignored if --config-file is specified)  [default: 300000]
	</tr>
	<tr>
		<td>--log-level [level]</td><td>minimum logging level (ignored if --config-file is specified)  [trace, debug, info, warn, error]
	</tr>
	<tr>
		<td>--max-cycles [size]</td><td>The maximum number of cycles to allow before throwing an exception  [default: 200001]
	</tr>
	<tr>
		<td>--max-loop-iterations [iterations]</td><td>the maximum number of iterations a loop can iterate before falling back to an unknown evaluation (ignored if --config-file is specified)  [default: 200000]
	</tr>
	<tr>
		<td>--max-recursion-limit [recursion limit]</td><td>the maximum recursion depth to evaluate before throwing a RangeError exception (ignored if --config-file is specified)  [default: 500]
	</tr>
	<tr>
		<td>-o, --output [format]</td><td>output format  [report, json, stream]
	</tr>
	<tr>
		<td>-p, --platform [platform]</td><td>the name of the OS being built-for, reflected in code via Ti.Platform.osname (ignored if --config-file is specified)
	</tr>
	<tr>
		<td>--plugins [plugins]</td><td>a comma separated list of plugin names to load (ignored if --config-file is specified)
	</tr>
	<tr>
		<td>-d, --project-dir [value]</td><td>the directory containing the project, otherwise the current working directory (ignored if --config-file is specified)
	</tr>
	<tr>
		<td>-R, --results-dir [value]</td><td>the path to the directory that will contain the generated results pages (ignored if --config-file is specified)
	</tr>
</table>

### Stream Output Format

```
titanium-code-processor subprocess <path/to/config/file>
```
The stream output format allows for JSON data to be passed in packets as the code processor runs. Data is passed back and forth using a custom two-layer packet format. The lower layer is completely custom, while the upper layer is formatted JSON and is encapsulated by the lower layer

#### Low Level Packet Format

The low level packet consists of four comma separated fields that forms a message
```
[Message Type],[Sequence ID],[Message Length],[data]
```
Note: the packet header at this level is ASCII formatted, although the data can theoretically be in any format

<table>
	<tr>
		<th>Name</th><th>Description</th>
	</tr>
	<tr>
		<td>MessageType</td><td>A three character sequence that currently is always 'REQ' (request)</td>
	</tr>
	<tr>
		<td>Sequence ID</td><td>A 32-bit, base 16 number that identifies the message. This value is always 8 characters long, and includes 0 padding if necessary.</td>
	</tr>
	<tr>
		<td>Message Length</td><td>A 32-bit, base 16 number that identifies the length of the message. This value is always 8 characters long, and includes 0 padding if necessary. Hex letters must be lower case.</td>
	</tr>
	<tr>
		<td>Data</td><td>The data for the message as specified in the [High Level Packet Format](#high-level-packet-format) section</td>
	</tr>
</table>

Example:

```
REQ,000079AC,0000000C,{foo: 'bar'}
```

#### High Level Packet Format

The high level packet is just a JSON string. The contents of the JSON object vary depending on message type and context.

A request always has the following definition:

* **messageType** _string_ A free form string defining the type of message, and is typically the name of an event
* **data** _any_ Any valid JSON value. Set to ```null``` if there is no data.

Example:

```JSON
{
	"messageType": "enteredFile",
	"data": {
		"filename": "path/to/file"
	}
}
```

#### Message Types

The code processor currently only sends one message, but others are planned for the future

**results**

The results from the project

* **errors** _array_ The errors from the project. The array is empty, but exists, if no errors were found
	* **type** _string_ The type of error
	* **description** _string_ A description of the error
	* **filename** _string_ The full path to the file where the error was detected
	* **line** _number_ The line number where the error was detected
	* **column** _number_ The column number where the error was detected
* **warnings** _array_ The warnings from the project. The array is empty, but exists, if no warnings were found
	* **type** _string_ The type of warning
	* **description** _string_ A description of the warning
	* **filename** _string_ The full path to the file where the warning was detected
	* **line** _number_ The line number where the warning was detected
	* **column** _number_ The column number where the warning was detected
* **plugins** _array_
	* **name** _string_ The name of the plugin, e.g. "ti-api-deprecation-finder"
	* **&lt;other key&gt;** Some other key specific to the plugin. See the plugin's README for detailed information
* **elapsedTime** _number_ The amount of time it took to process the project, in ms
* **resultsPath** _string_ The value of ```resultsPath``` passed to the code processor, for easy reference

Example:

```JSON
{
	"errors": [{
		"name": "SyntaxError",
		"description": "The description of the error",
		"data": {
			"otherKeys": "other data, including message, type, etc"
		},
		"filename": "path/to/file",
		"line": 0,
		"column": 0,
		"occurances": 0
	}],
	"warnings": [{
		"name": "SyntaxError",
		"description": "The description of the error",
		"data": {
			"otherKeys": "other data, including message, type, etc"
		},
		"filename": "path/to/file",
		"line": 0,
		"column": 0,
		"occurances": 0
	}],
	"plugins": [{
		"name": "plugin-name",
		"otherKeys": "other values"
	}],
	"elapsedTime": 0,
	"resultsPath": "resultsPath/from/config/file"
}
```

### Config File

The config file contains everything necessary for processing a project. Below is it's definition

* **sourceInformation** _object_ The path to the entry point
	 * **sourceDir** The directory containing the source code to be analyzed
	 * **entryPoint** The entry point for the project
	 * **projectDir** The project directory
	 * **originalSourceDir** &lt;optional&gt; The original directory that contained source code. Source maps map from here to sourceDir
	 * **sourceMaps** &lt;optional&gt; The source maps in key-value form. The key is a relative path to the file with sourceDir as the base, i.e. an absolute path to a file is at sourceDir + '/' + sourceMapKey

CHANGELOG.md
## 1.1.1

### Bug Fixes
* Fixed a bug where analyzer fails when only minimum Android SDK is installed.
* Fixed a bug to address the compatibility with dependent npm modules.

## 1.1.0

## New Features
* Implemented a proper CLI hook
* Unit tests now use RPC+Bonjour so that multiple systems can be combined to run unit tests faster
* Implemented support for properly analyzing skipped blocks, improving the accuracy of results considerably
* Added blacklisting of Alloy support files in coverage reports
* The supported platforms are now pulled from the SDK's manifest, meaning that it properly won't support iPhone on Windows
* Added support for recovering gracefully and quickly from infinite recursion

### Bug Fixes
* Fixed regressions introduced when the CLI was overhauled for 3.2.0
* Fixed a bug where explicitly setting an object's prototype to null caused the code processor to crash
* Fixed a bug where the this pointer was being set to module scope, not global scope, in certain situations
* Updated the unit test harness so that all tests run in global scope, fixing a bunch of unit tests
* Fixed a bug where eval, parseInt, etc had a prototype when they weren't supposed to.
* Fixed a bug where passing a hex string into parseInt was being parsed as decimal
* Fixed a bug with decode/encode URI methods not handling exceptions properly
* Fixed a bug where explicitly passing null or undefined to new Object() improperly threw an exception
* Fixed a bug where the builtin global constructors had their prototype defined twice
* Fixed a bug where strings were not handling numerical property access correctly
* Fixed a bug where the built-in global object's properties enumerable flags weren't being set properly
* Fixed a bug where some properties on the RegExp prototype weren't being added
* Fixed a bug where Object.defineProperties tried to evaluate non-enumerable properties
* Fixed a crash when passing in an invalid module path via configuration files
* Changed infinite loop detection to throw an error, not a warning

### Miscellany
* Merged the require-provider and common-globals plugins into the ti-api-provider
	* As a result, the code processor now only works with SDK 2.1.0 and newer

## 1.0.2

### Bug Fixes
* Created a work-around for the code processor exiting before it has printed all of the results to stdout

## 1.0.1

### New Features
* The code processor is now a CLI command! Run "titanium analyze --help" for more information
* Added support for Alloy projects
* Added support for the Tizen platform
* Lists of things in the results are now sorted alphabetically to more easily digest
* Reworked how source code information is specified in configuration files to allow for better flexibility
* Added the ability for plugins to elevate their summary into the main summary section

### Bug Fixes
* Fixed the case when a titanium object has an object property (typically a pseduo-type) from getting it's api listed twice in the API usage finder
* Reworked how plugins are initialized so that plugin names are available at initialization time
* Fixed a bug with native methods not creating their contexts properly
* Fixed a crash when calling Function.apply with an unknown value for arguments
* Fixed a crash when calling Array.indexOf and Array.unshift with an unknown this value
* Fixed a bug where running against the iPad platform didn't load iPhone modules

## 1.0.0

### New Features
* Added theme support for HTML results page
* Switched the CLI output to use the HTML templates
* Added runtime options and date/time finished to the results page
* Merged unknown-ambiguous-visualizer and unknown-callback-detector plugins
* Cleaned up the config file and output data formats. See the README for details
* Added more debug output
* Changed the default value of executionTimeLimit to 5 minutes
* Added generic cycle detection, of which infinite loops and infinite recursion are sub-classes of
* Added call tracing to visualization plugins
* Moved visualization plugins to enterprise-only

### Bug Fixes
* Fixed a few small bugs uncovered by the upgrade to node 0.10.0
* Updated the README's to contain accurate information about the Code Processor
* Fixed a bug with sending the results message in subprocess before the HTML results were ready
* Reworked invokeMethods=false internally to provide more accurate results by allowing native methods to still be called (console.log, Titanium methods, etc);
* Cleaned up undefined reference error messages
* Lots of small bug fixes
* The plugins sub-command now properly accepts paths
* Syntax errors now have filename/line/column information
* Syntax errors in a module required elsewhere no longer causes the code processor to crash
* Callbacks assigned to properties, such as Ti.Network.HTTPClient.onload, are now processed property
* Ti.Filesystem.ResourcesDirectory now returns something sensible
* Returns inside ambiguous blocks now behave properly
* Fixed a bug with node states being both skipped and visited
* Fixed unknown values short-circuiting expressions
* Fixed setting up the context properly for Ti.include calls

## 0.3.2 (28 February 2013)

### New Features
* Modified the Ti API usage finder to report both global api usage numbers and per-file api usage numbers
* Added HTML-bases results pages that studio can use to display results graphically
* Rearchitected unit tests so that they run 30 times faster

### Bug Fixes
* Fixed some more bugs surrounding setting up contexts for deferred methods
* Fixed a few miscellaneous bugs and regressions

lib/Runtime.js
/**
 * <p>Copyright (c) 2009-2013 by Appcelerator, Inc. All Rights Reserved.
 * Please see the LICENSE file for information about licensing.</p>
 *
 * This module provides information on the current state of the VM, the ability to change the state of the VM, and
 * communication-like services, such as logging and events.
 *
 * @module Runtime
 */

var path = require('path'),

	errors = [],
	warnings = [],

	logger,

	globalEvented,

	locationStack = [],

	processedFilesList = [],
	functions = [],

	asts = {},
	astIdentifier = 1,

	delayedFunctionsQueue = [],
	delayedFunctionsQueueIndex = 0;

/**
 * The global options to be used in processing
 *
 * @type Object
 * @name module:Runtime.options
 * @property {boolean} invokeMethods Indicates whether or not methods should be invoked
 * @property {boolean} evaluateLoops Indicates whether or not loops should be evaluated
 * @property {number} maxLoopIterations Indicates the maximum number of loop iterations to evaluate before erroring
 *		(infinite loop guard)
 * @property {bumber} maxRecursionLimit Indicates the maximum number of recurses to allow before claiming a recursive
 *		call is infinitely recursive (infinite recursion guard)
 * @property {number} cycleDetectionStackSize The size of the stack to use for cycle detection
 * @property {number} maxCycles The maximum number of cycles to allow
 * @property {boolean} logConsoleCalls If enabled, all console.* calls in a user's code are logged to the terminal
 * @property {(number | undefined)} executionTimeLimit Indicates the maximum time the app is allowed to run before erroring.
 *		Undefined means no limit
 * @property {boolean} exactMode Indicates whether or not the app should be evaluated in exact mode. Exact mode does not
 *		use ambiguous mode and throws an exception if an Unknown type is encountered
 * @property {boolean} processUnvisitedCode When set to true, all nodes and files that are not visited/skipped will be
 *		processed in ambiguous mode after all other code has been processed. While this will cause more of a project to
 *		be analyzed, this will decrease accuracy and can generate a lot of false positives.
 */
exports.options = {
	invokeMethods: true,
	evaluateLoops: true,
	maxLoopIterations: 200000,
	maxRecursionLimit: 10,
	cycleDetectionStackSize: 10000,
	maxCycles: 200001,
	logConsoleCalls: true,
	nativeExceptionRecovery: true,
	executionTimeLimit: 300000, // 5 minute timeout
	exactMode: false,
	processUnvisitedCode: false
};

/**
 * The wall-clock time when execution will time out, in unix epoch milliseconds
 *
 * @type number
 * @name module:Runtime.executionTimeLimit
 */
exports.executionTimeLimit = 0;

/**
 * A VM object representation of an error thrown from {@link module:base.throwNativeException} or {@link module:base/exception.throwException}
 *
 * @type module:base/types/object.ObjectType
 * @private
 */
exports._exception = undefined;

/**
 * The plugins associated with this run
 *
 * @type Object
 * @name module:Runtime.plugins
 */
exports.plugins = undefined;

/**
 * The source information
 *
 * @type module:CodeProcessor.sourceInformation
 * @name module:Runtime.sourceInformation
 */
exports.sourceInformation = undefined;

/**
 * Represents the location of an AST node in source code
 *
 * @typedef {Object} module:Runtime.location
 * @property {string} filename The name of the file that the node is in
 * @property {number} line The line number of the node
 * @property {number} column The column number of the node
 */
/**
 * Gets the current location being processed
 *
 * @method module:Runtime.getCurrentLocation
 * @return {module:Runtime.location} The current location
 */
exports.getCurrentLocation = getCurrentLocation;
function getCurrentLocation() {
	return locationStack[locationStack.length - 1] || {
		filename: '',
		line: 0,
		column: 0
	};
}

/**
 * Sets the current location
 *
 * @method module:Runtime.setCurrentLocation
 * @param {string} filename The filename
 * @param {number} line The current line
 * @param {number} column The current column
 */
exports.setCurrentLocation = setCurrentLocation;
function setCurrentLocation(filename, line, column) {
	locationStack.push({
		filename: filename,
		line: line,
		column: column
	});
}

/**
 * Maps a location to its original source
 *
 * @method module:Runtime.mapLocation
 * @param {module:Runtime.location} location The location to map
 * @return {module:Runtime.location} The mapped location, if it can be mapped, else the original location
 */
exports.mapLocation = mapLocation;
function mapLocation(location) {
	var sourceInformation = exports.sourceInformation,
		sourceMaps = sourceInformation && sourceInformation.sourceMaps,
		sourceMap,
		mappedLocation;
	if (sourceMaps) {
		sourceMap = sourceMaps[path.relative(exports.sourceInformation.sourceDir, location.filename)];
		if (sourceMap) {
			mappedLocation = sourceMap.originalPositionFor({
				line: location.line,
				column: location.column
			});
			return {
				filename: path.join(mappedLocation.source),
				line: mappedLocation.line,
				column: mappedLocation.column
			};
		} else {
			return location;
		}
	} else {
		return location;
	}
}

/**
 * Exits the current location, from a runtime perspective
 *
 * @method module:Runtime.exitCurrentLocation
 */
exports.exitCurrentLocation = exitCurrentLocation;
function exitCurrentLocation() {
	locationStack.pop();
}

/**
 * Checks if the current location's file is blacklisted or not
 *
 * @method module:Runtime.isCurrentFileBlacklisted
 * @return {boolean} Whether or not the current file has been blacklisted
 */
exports.isCurrentFileBlacklisted = isCurrentFileBlacklisted;
function isCurrentFileBlacklisted() {
	return isFileBlacklisted(getCurrentLocation().filename);
}

/**
 * Checks if the given files has been blacklisted
 *
 * @method module:Runtime.isFileBlacklisted
 * @param  {string} filename The path to the file to check
 * @return {boolean} Whether or not the given file has been blacklisted
 */
exports.isFileBlacklisted = isFileBlacklisted;
function isFileBlacklisted(filename) {
	var i, len,
		whiteList = exports.options.whiteList;
	if (whiteList) {
		for (i = 0, len = whiteList.length; i < len; i++) {
			if (filename.indexOf(whiteList[i]) === 0) {
				return false;
			}
		}
		return true;
	}
	return false;
}

/**
 * Adds an AST to the set of processed ASTs
 *
 * @method module:Runtime.setAST
 * @param {module:AST.node} ast The ast node to set
 * @param {string} [filename] The name of the file that generated this AST. If not supplied, an identifier is automatically generated
 */
exports.setAST = setAST;
function setAST(ast, filename) {
	if (!filename) {
		filename = '@unnamed_ast_' + (astIdentifier++);
	}
	asts[filename] = ast;
}

/**
 * Gets the list of ASTs that have been processed
 *
 * @method module:Runtime.getASTSet
 * @return {Object} A dictionary of the ASTs processed, with the filename/identifier as the key and the AST as the value
 */
exports.getASTSet = getASTSet;
function getASTSet() {
	return asts;
}

/**
 * Gets the list of files that have been processed so far
 *
 * @method module:Runtime.getProcessedFilesList
 * @return {Array.<string>} An array of filenames
 */
exports.getProcessedFilesList = getProcessedFilesList;
function getProcessedFilesList() {
	return processedFilesList;
}

/**
 * Gets the list of files that have not been processed so far (and may not be ever)
 *
 * @method module:Runtime.getUnprocessedFilesList
 * @return {Array.<string>} An array of filenames
 */
exports.getUnprocessedFilesList = getUnprocessedFilesList;
function getUnprocessedFilesList() {
	var i, len,
		unprocessedFiles = [];
	for (i = 0, len = exports.fileList.length; i < len; i++) {
		if (processedFilesList.indexOf(exports.fileList[i]) === -1) {
			unprocessedFiles.push(exports.fileList[i]);
		}
	}
	return unprocessedFiles;
}

/**
 * Adds a function to the list of all functions
 *
 * @method module:Runtime.addFunction
 * @param {module:AST.node} func The function to add
 */
exports.addFunction = addFunction;
function addFunction (func) {
	if (functions.indexOf(func) === -1) {
		functions.push(func);
	}
}

/**
 * Gets all functions that haven't been processed yet
 *
 * @method module:Runtime.getNextUnprocessedFunction
 * @return {Array.<module:AST.node>} The list of unprocessed functions
 */
exports.getNextUnprocessedFunction = getNextUnprocessedFunction;
function getNextUnprocessedFunction() {
	var bodyNode,
		i, len = functions.length;
	for (i = 0; i < len; i++) {
		bodyNode = functions[i].body[0];
		if (bodyNode && !bodyNode._visited && !bodyNode._skipped) {
			return functions[i];
		}
	}
}

/**
 * Queues a function for later evaluation
 *
 * @method module:Runtime.queueFunction
 * @param {module:base/types/function.FunctionType} func The function to execute later
 * @param {(module:base.BaseType | undefined)} thisVal The this value
 * @param {Array.<module:base.BaseType>} args The function arguments
 * @param {boolean} ambiguousContext Whether or not this should be evaluated in ambiguous mode
 * @param {boolean} skippedContext Whether or not this should be evaluated in skipped mode
 */
exports.queueFunction = queueFunction;
function queueFunction(func, thisVal, args, ambiguousContext, skippedContext) {

	// Make sure that the function isn't already in the queue
	var location = getCurrentLocation(),
		filename = location.filename,
		line = location.line,
		column = location.column,
		i, len;
	for (i = 0, len = delayedFunctionsQueue.length; i < len; i++) {
		if (delayedFunctionsQueue[i].filename === filename &&
				delayedFunctionsQueue[i].line === line &&
				delayedFunctionsQueue[i].column === column) {
			return;
		}
	}

	// Set the thisVal to its default here since we won't be able to get it later
	if (!func.strict && (!thisVal || thisVal.className === 'Null' || thisVal.className === 'Undefined')) {
		// I really hate doing an inline require, but it's necessary to break a circular dependency without moving
		// everything in this file to Context.js
		thisVal = require('./Base').getModuleContext().thisBinding;
	}

	// Queue the function
	delayedFunctionsQueue.push({
		func: func,
		thisVal: thisVal,
		args: args || [],
		ambiguousContext: !!ambiguousContext,
		skippedContext: !!skippedContext,
		filename: filename,
		line: line,
		column: column
	});
}

/**
 * Gets the next queued function to be executed
 *
 * @method module:Runtime.getNextQueuedFunction
 * @return {module:base/types/function.FunctionType} The next function to be executed
 */
exports.getNextQueuedFunction = getNextQueuedFunction;
function getNextQueuedFunction() {
	if (delayedFunctionsQueue[delayedFunctionsQueueIndex]) {
		return delayedFunctionsQueue[delayedFunctionsQueueIndex++];
	}
}

/**
 * @private
 */
function createEventObject(type, description, data) {
	var location = getCurrentLocation(),
		result = {
			type: type,
			description: description,
			data: data
		};
	if (data && data.exception) {
		result.filename = data.exception.filename;
		result.line = data.exception.line;
		result.column = data.exception.column;
	} else if (location) {
		result.filename = location.filename;
		result.line = location.line;
		result.column = location.column;
	}

	return result;
}

// ******** Evented Object Methods ********
/**
 * Base class for adding event support to other objects.
 *
 * @constructor module:Runtime.Evented
 */
exports.Evented = Evented;
function Evented() {
	this.listeners = {};
}

/**
 * An event object
 *
 * @typedef {Object} module:Runtime.eventObject
 * @property {string} type The type of the event
 * @property {string} description The description of the event
 * @property {(string | undefined)} filename The name of the file being processed when the event was triggered, if any
 * @property {(number | undefined)} line The line number of the file being processed when the event was triggered, if any
 * @property {(number | undefined)} column The column number of the file being processed when the event was triggered, if any
 * @property {Object} data The event-specific information
 */
/**
 * A callback to be called when an event is fired
 *
 * @callback module:Runtime.eventedOnCallback
 * @param {module:Runtime.eventObject} e The event information
 */
/**
 * Adds an event listener for the given event
 *
 * @method module:Runtime.Evented#on
 * @param {string} name The name of the event to listen to, e.g. 'parseError'
 * @param {module:Runtime.eventedOnCallback} callback The function to call when the event is fired
 */
Evented.prototype.on = function on(name, callback) {
	var eventListeners = this.listeners;
	if (!eventListeners[name]) {
		eventListeners[name] = [];
	}
	eventListeners[name].push(callback);
};

/**
 * Fires an event to the listener set
 *
 * @method module:Runtime.Evented#fireEvent
 * @param {string} name The name of the event, e.g. 'projectProcessingEnd'
 * @param {string} description A description of the event, e.g. 'Processing completed successfully'
 * @param {Object} [data] Exra data associated with the event
 */
Evented.prototype.fireEvent = function fireEvent(name, description, data) {
	var listeners = this.listeners[name],
		i, len = listeners ? listeners.length : 0,
		e = createEventObject(name, description, data);
	log('trace', 'Firing event "' + name + '": ' + description + (e.filename ? ' (' + e.filename + ':' + e.line + ')' : ''));
	for (i = 0; i < len; i++) {
		listeners[i](e);
	}
};

// ******** Global Event Methods ********

globalEvented = new Evented();

/**
 * Adds an event listener for the given event
 *
 * @method module:Runtime.on
 * @param {string} name The name of the event to listen to, e.g. 'parseError'
 * @param {module:Runtime.eventedOnCallback} callback The function to call when the event is fired
 */
exports.on = on;
function on() {
	return globalEvented.on.apply(globalEvented, arguments);
}

/**
 * Fires an event to the global listener set
 *
 * @method module:Runtime.fireEvent
 * @param {string} name The name of the event, e.g. 'projectProcessingEnd'
 * @param {string} description A description of the event, e.g. 'Processing completed successfully'
 * @param {Object} [data] Exra data associated with the event
 */
exports.fireEvent = fireEvent;
function fireEvent() {
	return globalEvented.fireEvent.apply(globalEvented, arguments);
}

// ******** Error/warning Methods ********

/**
 * Reports an error
 *
 * @method module:Runtime.reportError
 * @param {string} type The type of the error, e.g. 'SyntaxError'
 * @param {string} description A description of the error.
 * @param {Array.<module:Runtime.location>} stackTrace The current stack trace
 */
exports.reportError = reportError;
function reportError(type, description, stackTrace) {

	// Squash the reporting of the error if we are in skipped mode
	if (require('./Base').isSkippedMode()) {
		return;
	}

	var i, len,
		error,
		eventObject = createEventObject(type, description || '');
	for (i = 0, len = errors.length; i < len; i++) {
		error = errors[i];
		if (error.file === eventObject.file && error.line === eventObject.line && error.column === eventObject.column &&
				error.type === eventObject.type && error.description === eventObject.description) {
			error.occurances++;
			return;
		}
	}
	eventObject.occurances = 1;
	fireEvent('errorReported', type + ': ' + description, {
		type: type,
		description: description
	});
	log('error', type + ': ' + description +
		(stackTrace ? '\n    at ' + stackTrace.join('\n    at ') : ''));
	errors.push(eventObject);
}

/**
 * Reports a syntax error from uglify. Syntax errors from Uglify are handled differently because the code processor
 * needs to get the location via a different means than all other errors since they aren't generated internally.
 *
 * @method module:Runtime.reportError
 * @param {Object} uglifyError The exception thrown by Uglify
 */
exports.reportUglifyError = reportUglifyError;
function reportUglifyError(uglifyError) {

	// Squash the reporting of the error if we are in skipped mode
	if (require('./Base').isSkippedMode()) {
		return;
	}

	var i, len,
		error,
		eventObject = createEventObject('SyntaxError', uglifyError.message || '');
	eventObject.filename = uglifyError.filename;
	eventObject.line = uglifyError.line;
	eventObject.column = uglifyError.column;
	for (i = 0, len = errors.length; i < len; i++) {
		error = errors[i];
		if (error.file === eventObject.file && error.line === eventObject.line && error.column === eventObject.column &&
				error.type === eventObject.type && error.description === eventObject.description) {
			error.occurances++;
			return;
		}
	}
	eventObject.occurances = 1;
	fireEvent('errorReported', 'SyntaxError: ' + uglifyError.message, {
		type: 'SyntaxError',
		description: uglifyError.message
	});
	log('error', 'SyntaxError: ' + uglifyError.message);
	errors.push(eventObject);
}

/**
 * Reports a warning
 *
 * @method module:Runtime.reportWarning
 * @param {string} type The type of the warning, e.g. 'requireMissing.'  Note: the type will be made lowercase.
 * @param {string} description A description of the warning, e.g. 'A required file could not be found'
 */
exports.reportWarning = reportWarning;
function reportWarning(type, description) {

	// Squash the reporting of the error if we are in skipped mode
	if (require('./Base').isSkippedMode()) {
		return;
	}

	var i, len,
		warning,
		eventObject = createEventObject(type, description || '');
	for (i = 0, len = warnings.length; i < len; i++) {
		warning = warnings[i];
		if (warning.file === eventObject.file && warning.line === eventObject.line && warning.column === eventObject.column &&
				warning.type === eventObject.type && warning.description === eventObject.description) {
			warning.occurances++;
			return;
		}
	}
	eventObject.occurances = 1;
	fireEvent('warningReported', type + ': ' + description, {
		type: type,
		description: description
	});
	log('warn', type + ': ' + description);
	warnings.push(eventObject);
}

/**
 * Gets the list of all errors reported via {@link module:Runtime.reportError}.
 *
 * @method module:Runtime.getReportedErrors
 * @return {Array.<module:Runtime.eventObject>} The list of errors
 */
exports.getReportedErrors = getReportedErrors;
function getReportedErrors() {
	return errors;
}

/**
 * Gets the list of all warnings reported via {@link module:Runtime.reportWarning}.
 *
 * @method module:Runtime.getReportedWarnings
 * @return {Array.<module:Runtime.eventObject>} The list of warnings
 */
exports.getReportedWarnings = getReportedWarnings;
function getReportedWarnings() {
	return warnings;
}

// ******** Logging Methods ********

/**
 * Sets the logger that everything will use
 *
 * @method module:Runtime.setLogger
 * @param {Object} newLogger The Winston logger to set
 */
exports.setLogger = setLogger;
function setLogger(newLogger) {
	logger = newLogger;
}

/**
 * Logs a message
 *
 * @method module:Runtime.log
 * @param {string} level The log level of the message
 * @param {string} message The message to log
 */
exports.log = log;
function log(level, message) {
	if (logger) {
		logger.log(level, message);
	}
}

// ******** Runtime events to listen to ********

/**
 * @private
 */
on('enteredFile', function(e) {
	processedFilesList.push(e.data.filename);
});